services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,202 @@
+import javax.annotation.PostConstruct;
+
+/**
+ * Implementation of {@link MessagingChannel} for kafka.
+ */
+@Component
+public class KafkaMessagingChannel implements MessagingChannel {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaMessagingChannel.class);
+
+    @Value("${northbound.messages.expiration.minutes}")
+    private int expiredTime;
+
+    /**
+     * Requests that are in progress of processing.
+     */
+    private Map<String, CompletableFuture<InfoData>> pendingRequests;
+    private Map<String, CompletableFuture<List<InfoData>>> pendingChunkedRequests;
+
+    /**
+     * Collects messages that are not related to any known chain.
+     */
+    private Map<String, ChunkedInfoMessage> unlinkedChunks;
+
+    /**
+     * Chains of chunked messages, it is filling by messages one by one as soon as the next linked message is received.
+     */
+    private Map<String, List<ChunkedInfoMessage>> messagesChains;
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * Creates storages, that are able to remove outdated messages and requests.
+     */
+    @PostConstruct
+    public void setUp() {
+        pendingRequests = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new ConcurrentHashMap<>());
+        pendingChunkedRequests = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new ConcurrentHashMap<>());
+
+        unlinkedChunks = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+        messagesChains = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoData> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+        pendingRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<InfoData>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<InfoData>> future = new CompletableFuture<>();
+        pendingChunkedRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
```

Comment:
It's not clear how to handle the timeout case: a future object with no response must be automatically removed from pendingRequests and pendingChunkedRequests.

Reply:
Fixed. In case of timeout DeferredResult.onTimeout will be called by spring (see org.openkilda.northbound.utils.async.CompletableFutureReturnValueHandler.ControlledDeferredResult)